File: Digito Verificador/Modulo 11/DV Verificador/verifdig_mod11.pas
program ValidarMatricula;

uses
  SysUtils;

function ValidarMatricula(const MATRICULA: string): Boolean;
var
  I: Integer;
begin
  if Length(MATRICULA) <> 9 then
    Exit(False);
  for I := 1 to 3 do
    if not (MATRICULA[I] in ['0'..'9']) then
      Exit(False);
  if MATRICULA[4] <> '.' then
    Exit(False);
  for I := 5 to 7 do
    if not (MATRICULA[I] in ['0'..'9']) then
      Exit(False);
  if MATRICULA[8] <> '-' then
    Exit(False);
  if not (MATRICULA[9] in ['0'..'9']) then
    Exit(False);
  ValidarMatricula := True;
end;

procedure RemoverCaractere(var CADEIA: string; CARACTERE: Char);
var
  POS1, POS2: Integer;
begin
  POS1 := 1;
  POS2 := 1;
  while POS1 <= Length(CADEIA) do
  begin
    if CADEIA[POS1] <> CARACTERE then
    begin
      CADEIA[POS2] := CADEIA[POS1];
      POS2 := POS2 + 1;
    end;
    POS1 := POS1 + 1;
  end;
  SetLength(CADEIA, POS2 - 1);
end;

function Mod11Check(const MATRICULA: string): Boolean;
var
  MATRIC_INT: array[1..7] of Integer;
  I, SOMA: Integer;
begin
  for I := 1 to 7 do
    MATRIC_INT[I] := Ord(MATRICULA[I]) - 48; // ASCII 48 = "0"
  SOMA := 0;
  for I := 1 to 7 do
    SOMA := SOMA + MATRIC_INT[I] * (9 - I);
  Mod11Check := (SOMA mod 11) = 0;
end;

var
  MATRICULA: string;
  DVC: Boolean;
  DVE: Char;

begin
  repeat
    Write('Informe matricula no formato 999.999-9: ');
    ReadLn(MATRICULA);
    if not ValidarMatricula(MATRICULA) then
    begin
      WriteLn;
      WriteLn('Matricula invalida. Tente novamente.');
      WriteLn;
    end;
  until ValidarMatricula(MATRICULA);

  DVC := Mod11Check(MATRICULA);
  DVE := MATRICULA[10];
  WriteLn;
  if (DVC and (DVE = '0')) or (not DVC and (DVE <> '0')) then
    WriteLn('Numero de matricula valido')
  else
    WriteLn('Numero de matricula invalido');
  WriteLn;
  Write('Pressione Enter para encerrar...');
  ReadLn;
end.

File: Digito Verificador/Modulo 11 Padrao/DV Verificador/verifdig_mod11.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func isDigit(CARACTERE byte) bool {
	return CARACTERE >= '0' && CARACTERE <= '9'
}

func validarMatricula(MATRICULA string) bool {
	if len(MATRICULA) != 9 {
		return false
	}
	for I := 0; I <= 2; I++ {
		if !isDigit(MATRICULA[I]) {
			return false
		}
	}
	if MATRICULA[3] != '.' {
		return false
	}
	for I := 4; I <= 6; I++ {
		if !isDigit(MATRICULA[I]) {
			return false
		}
	}
	if MATRICULA[7] != '-' {
		return false
	}
	if !isDigit(MATRICULA[8]) {
		return false
	}
	return true
}

func removerCaractere(CADEIA string, CARACTERE byte) string {
	NOVA_CADEIA := strings.Builder{}
	for I := 0; I < len(CADEIA); I++ {
		if CADEIA[I] != CARACTERE {
			NOVA_CADEIA.WriteByte(CADEIA[I])
		}
	}
	return NOVA_CADEIA.String()
}

func mod11Check(MATRICULA string) int {
	MATRIC_INT := make([]int, 7)
	for I := 0; I <= 6; I++ {
		MATRIC_INT[I], _ = strconv.Atoi(string(MATRICULA[I]))
	}
	SOMA := 0
	for I := 0; I <= 5; I++ {
		SOMA += MATRIC_INT[I] * (7 - I)
	}
	DV := (11 - SOMA % 11) % 10
	return DV
}

func main() {
	var MATRIC_CPT string
	var MATRIC_SPT string
	var DVC, DVE int

	for {
		fmt.Print("Informe matricula no formato 999.999-9: ")
		fmt.Scanln(&MATRIC_CPT)
		if !validarMatricula(MATRIC_CPT) {
			fmt.Println("\nMatricula invalida. Tente novamente.\n")
		} else {
			break
		}
	}

	MATRIC_SPT = MATRIC_CPT
	MATRIC_SPT = removerCaractere(MATRIC_SPT, '.')
	MATRIC_SPT = removerCaractere(MATRIC_SPT, '-')
	DVC = mod11Check(MATRIC_SPT)
	MATRIC_INT := make([]int, 7)
	for I := 0; I <= 5; I++ {
		MATRIC_INT[I], _ = strconv.Atoi(string(MATRIC_SPT[I]))
	}
	DVE, _ = strconv.Atoi(string(MATRIC_SPT[6]))
	fmt.Println()
	if DVC == DVE {
		fmt.Println("Numero de matricula valido")
	} else {
		fmt.Println("Numero de matricula invalido")
	}
	fmt.Println()
	fmt.Print("Tecle <Enter> para encerrar...")
	fmt.Scanln()
}
